### PR TITLE
Implement glClipPlane and partial GL_LIGHTING emulation

### PR DESCRIPTION
**Pull request by rbrune**
To get some rather old code base running with the opengl emulation system I started working on adding some missing features. The glClipPlane functionality being first.

### CODE REVIEW

File: src/library_glemu.js [hunk 1]
```diff
@@ -3415,9 +3489,27 @@ var LibraryGLEmulation = {
...
+      GLImmediate.matrixLib.mat4.inverse(tmpMV);
+      GLImmediate.matrixLib.mat4.transpose(tmpMV);
+      GLImmediate.matrixLib.mat4.multiplyVec4(tmpMV, GLEmulation.clipPlaneEquation[clipPlaneId]);
+    }
+  },
+
+  glLightfv: function() {}, //{ throw 'glLightfv: TODO' },
+  glLightModelf: function() {}, //{ throw 'glLightModelf: TODO' },
+  glLightModelfv: function() {}, //{ throw 'glLightModelfv: TODO' },
+  glMaterialfv: function() {}, //{ throw 'glMaterialfv: TODO' },
```
Comment (kripken) [inline src/library_glemu.js:3512]: it would be good to warn here at least, when in `ASSERTIONS`, that these are not fully implemented.

**Review by kripken — COMMENTED** [review]
Thanks for the PR!

This kind of thing looks good to me. I'm not a GL expert so I can't be sure of all the details, but it looks generally right, there is no risk of regressing existing code, and if it works on your codebase then that is a good.

One comment below, and also, we should add a test for this. The test would go in `tests/test_browser.py`, can look for `LEGACY_GL_EMULATION` to find similar tests.

File: src/library_glemu.js (hunk 1 above)
Comment (rbrune) [reply src/library_glemu.js]: Yes. That was not supposed to be in the initial PR. See comment below.

**Comment by rbrune** [comment]
Now I added functionality to support GL_LIGHTING emulation.
Partially (no light attenuation modes and only directional lights) but enough to get a lot of stuff working. (And enough to accurately emulate the behaviour of the ancient code base I'm making buildable with emscripten).

Regarding tests. Is the policy to add one for each functionality? E.g. would be one for clip planes and one for a scene with lights enough?

**Comment by kripken** [comment]
For things like this I think a single test that covers multiple related things is good enough, assuming it's easier to write that way.

(In general more fine-grained tests are better, but browser tests are slower, and the GL emulation code already has many such multiple tests.)

**Comment by rbrune** [comment]
Thanks!

A test including a reference rendering png has been added. It's not really an exhaustive test as it only uses a single clip plane and a single light and material. The code base I developed the emulation features for does use multiple lights and materials in various combinations.

File: src/library_glemu.js [hunk 2]
```diff
@@ -3415,9 +3652,99 @@ var LibraryGLEmulation = {
...
+      // apply inverse transposed current modelview matrix when setting clip plane
+      var tmpMV = GLImmediate.matrixLib.mat4.create(GLImmediate.matrix[0]);
+      GLImmediate.matrixLib.mat4.inverse(tmpMV);
+      GLImmediate.matrixLib.mat4.transpose(tmpMV);
+      GLImmediate.matrixLib.mat4.multiplyVec4(tmpMV, GLEmulation.clipPlaneEquation[clipPlaneId]);
+    }
+  },
+
+  glLightfv: function(light, pname, param) {
+      if ((light >= 0x4000) && (light < 0x4008)  /* GL_LIGHT0 to GL_LIGHT7 */) {
```
Comment (kripken) [inline src/library_glemu.js:3673]: indentation here and in other places looks incorrect - it should always be 2 space (not 4, and not tabs).

**Review by kripken — COMMENTED** [review]
Thanks! Looks good aside from whitespace.

File: src/library_glemu.js (hunk 2 above)
Comment (rbrune) [reply src/library_glemu.js]: Sorry for that. Fixed it!

File: src/library_glemu.js [hunk 3]
```diff
@@ -175,6 +241,26 @@ var LibraryGLEmulation = {
...
+          if (GLEmulation.lightEnabled[lightId] != false) {
+            GLImmediate.currentRenderer = null; // light parameter is part of the FFP shader state, we must re-lookup the renderer to use.
+            GLEmulation.lightEnabled[lightId] = false;
+          }
+          return;
+        } else if (cap == 0xB50 /* GL_LIGHTING */) {
+          if (GLEmulation.lightingEnabled != false) {
+            GLImmediate.currentRenderer = null; // Fog parameter is part of the FFP shader state, we must re-lookup the renderer to use.
+            GLEmulation.lightingEnabled = false;
+        }
```
Comment (kripken) [inline src/library_glemu.js:262]: something looks wrong here and on the line below with the indentation.

Comment (rbrune) [reply src/library_glemu.js]: Embarrassing. Fixed that too.

**Review by kripken — APPROVED** [review]
Thanks!

File: src/library_glemu.js [hunk 4]
```diff
@@ -2112,6 +2230,60 @@ var LibraryGLEmulation = {
...
+                vsClipPlaneDefs += 'uniform vec4 u_clipPlaneEquation' + clipPlaneId + ';';
+                vsClipPlaneDefs += 'varying float v_clipDistance' + clipPlaneId + ';';
+                vsClipPlaneInit += '  v_clipDistance' + clipPlaneId + ' = dot(ecPosition, u_clipPlaneEquation' + clipPlaneId + ');';
+                fsClipPlaneDefs += 'varying float v_clipDistance' + clipPlaneId + ';';
+                fsClipPlanePass += '  if(v_clipDistance' + clipPlaneId + ' < 0.0) discard;';
+              }
+            }
+
+            vsLightingDefs = '';
+            vsLightingPass = '';
```
Comment (kripken) [inline src/library_glemu.js:2248]: I think the CI errors should be fixed by making these two `var` declarations.

**Comment by kripken** [comment]
Tests pass!

**Pull request merged**